File: solutions/erlang/rotational-cipher/1/src/rotational_cipher.erl
-module(rotational_cipher).
-export([encrypt/2, decrypt/2, test_version/0]).

encrypt(Text, Rot) ->
  [rotate(C, Rot) || C <- Text].

decrypt(Text, Rot) ->
  [rotate(C, (26 - Rot)) || C <- Text].

rotate(C, Rot) when $a =< C andalso C =< $z ->
  ((C - $a) + Rot) rem 26 + $a;
rotate(C, Rot) when $A =< C andalso C =< $Z ->
  ((C - $A) + Rot) rem 26 + $A;
rotate(C, _Rot) ->
  C.

test_version() -> 1.

File: solutions/elixir/rotational-cipher/2/lib/rotational_cipher.ex
defmodule RotationalCipher do
  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """
  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift) do
    text |> String.to_charlist() |> Enum.map(&rot(&1, shift)) |> to_string()
  end

  defp rot(ch, shift) when ch in ?A..?Z, do: ?A + rem(ch - ?A + shift, 26)
  defp rot(ch, shift) when ch in ?a..?z, do: ?a + rem(ch - ?a + shift, 26)
  defp rot(ch, _), do: ch
end
